File: server/report-generator.ts
/**
 * Report Generator
 *
 * This module handles background generation of Product Level Reports.
 * All data fetching and processing is done in Node.js via report-worker.ts.
 *
 * Architecture:
 *   Node.js: Task scheduling, status management, API routing, notifications,
 *            Facebook API data fetching, data processing, S3 upload, catalog updates
 *   Communication: Worker updates progress directly in the database (batch_jobs table)
 */

import {
  updateBatchJob,
  createSavedReport,
  updateSavedReport,
} from "./db";
import { notifyOwner } from "./_core/notification";
import { BatchJob } from "../drizzle/schema";
import { runReportWorker, WorkerConfig, WorkerResult } from "./report-worker";

interface ReportConfig {
  adAccountId?: string;
  accessToken?: string;
  dateStart?: string;
  dateEnd?: string;
  level?: string;
  breakdown?: string;
  minSpend?: string;
  minCTR?: string;
  dateRangeType?: string;
  // For combined workflow (report + catalog update)
  updateToCatalog?: boolean;
  catalogId?: string;
  catalogAccessToken?: string;
  customLabel4?: string;
  enableCustomLabel4?: boolean;
  // Custom number fields (0-4)
  customNumbers?: Record<string, string>;
  // Custom label fields (0-4)
  customLabels?: Record<string, string>;
  // Schedule tracking
  scheduleId?: number;
  scheduleName?: string;
  scheduleRunId?: number;
  configIndex?: number;
  configName?: string;
}

/**
 * Calculate date range based on type
 */
function calculateDateRange(dateRangeType: string): { dateStart: string; dateEnd: string } {
  const now = new Date();
  let dateStart: Date;
  let dateEnd: Date = new Date(now);
  dateEnd.setDate(dateEnd.getDate() - 1); // Yesterday

  switch (dateRangeType) {
    case 'last_7_days':
      dateStart = new Date(now);
      dateStart.setDate(dateStart.getDate() - 7);
      break;
    case 'last_14_days':
      dateStart = new Date(now);
      dateStart.setDate(dateStart.getDate() - 14);
      break;
    case 'last_30_days':
      dateStart = new Date(now);
      dateStart.setDate(dateStart.getDate() - 30);
      break;
    case 'last_week': {
      // Last complete week (Monday to Sunday)
      const dayOfWeek = now.getDay();
      const daysToLastSunday = dayOfWeek === 0 ? 7 : dayOfWeek;
      dateEnd = new Date(now);
      dateEnd.setDate(dateEnd.getDate() - daysToLastSunday);
      dateStart = new Date(dateEnd);
      dateStart.setDate(dateStart.getDate() - 6);
      break;
    }
    case 'last_month':
      // Last complete month
      dateEnd = new Date(now.getFullYear(), now.getMonth(), 0); // Last day of previous month
      dateStart = new Date(now.getFullYear(), now.getMonth() - 1, 1); // First day of previous month
      break;
    default:
      // Default to last 7 days
      dateStart = new Date(now);
      dateStart.setDate(dateStart.getDate() - 7);
  }

  return {
    dateStart: dateStart.toISOString().split('T')[0],
    dateEnd: dateEnd.toISOString().split('T')[0],
  };
}

/**
 * Process a report generation job using the Node.js worker.
 *
 * Steps:
 *   1. Validate config and create saved report record
 *   2. Build worker config
 *   3. Run the Node.js report worker (async, same process)
 *   4. Handle result and send notifications
 */
export async function processReportGenerationJob(job: BatchJob, startTime: number): Promise<void> {
  const config = job.config as ReportConfig;

  // Validate required fields
  if (!config.adAccountId || !config.accessToken) {
    throw new Error('Missing required config fields: adAccountId or accessToken');
  }

  const adAccountId = config.adAccountId;
  const accessToken = config.accessToken;

  // Calculate date range
  let dateStart = config.dateStart;
  let dateEnd = config.dateEnd;

  if (config.dateRangeType && !dateStart) {
    const range = calculateDateRange(config.dateRangeType);
    dateStart = range.dateStart;
    dateEnd = range.dateEnd;
  }

  if (!dateStart || !dateEnd) {
    throw new Error('Missing date range configuration');
  }

  const level = config.level || 'account';
  const breakdown = config.breakdown || 'product_id';

  // Build filters
  const filters: Array<{ field: string; operator: string; value: any }> = [];
  if (config.minSpend) {
    filters.push({ field: 'spend', operator: 'GREATER_THAN', value: parseFloat(config.minSpend) });
  }
  if (config.minCTR) {
    filters.push({ field: 'inline_link_click_ctr', operator: 'GREATER_THAN', value: parseFloat(config.minCTR) });
  }

  // Create saved report record
  const reportId = await createSavedReport({
    userId: job.userId,
    name: `Report ${dateStart} to ${dateEnd}`,
    adAccountId,
    dateStart,
    dateEnd,
    level,
    breakdown,
    minSpend: config.minSpend,
    minCTR: config.minCTR,
    status: 'generating',
    source: 'manual',
  });

  if (!reportId) {
    throw new Error('Failed to create saved report record');
  }

  // Update job with report ID
  await updateBatchJob(job.id, {
    reportId,
    statusMessage: 'Starting report worker…',
  });

  // Build worker config
  const workerConfig: WorkerConfig = {
    jobId: job.id,
    reportId,
    userId: String(job.userId),
    adAccountId,
    accessToken,
    dateStart,
    dateEnd,
    level,
    breakdown,
    filters: filters.length > 0 ? filters : null,
    updateToCatalog: config.updateToCatalog || false,
    catalogId: config.catalogId,
    catalogAccessToken: config.catalogAccessToken,
    customLabel4: config.customLabel4,
    enableCustomLabel4: config.enableCustomLabel4,
    customNumbers: config.customNumbers,
    customLabels: config.customLabels,
    scheduleRunId: config.scheduleRunId,
  };

  console.log(`[ReportGenerator] Starting Node.js worker for job ${job.id}…`);

  // Run the Node.js worker directly (no Python spawn)
  const result: WorkerResult = await runReportWorker(workerConfig);

  if (result.success) {
    console.log(`[ReportGenerator] Worker completed successfully for job ${job.id}`);

    // Send notification to owner
    try {
      const durationMinutes = Math.round((result.durationMs || 0) / 60000);
      const notificationTitle = config.updateToCatalog
        ? `✅ Report + Catalog Update Completed`
        : `✅ Report Generation Completed`;
      const notificationContent = [
        `**Job ID:** ${job.id}`,
        `**Account:** ${config.adAccountId}`,
        `**Products:** ${(result.totalItems || 0).toLocaleString()}`,
        `**Duration:** ${durationMinutes} minutes`,
        config.updateToCatalog ? `**Catalog Updated:** Yes` : '',
        `\n[View Reports](/reports)`,
      ].filter(Boolean).join('\n');

      await notifyOwner({
        title: notificationTitle,
        content: notificationContent,
      });
      console.log(`[ReportGenerator] Notification sent for job ${job.id}`);
    } catch (notifyError) {
      console.warn(`[ReportGenerator] Failed to send notification:`, notifyError);
    }
  } else {
    // Worker reported failure
    const errorMsg = result.error || 'Unknown worker error';
    console.error(`[ReportGenerator] Worker failed for job ${job.id}: ${errorMsg}`);

    // Update report as failed
    await updateSavedReport(reportId, {
      status: 'failed',
      errorMessage: errorMsg.substring(0, 500),
    });

    throw new Error(errorMsg.substring(0, 500));
  }
}

File: backend/app/services/report_generator.py
"""
Report Generator - orchestrates report generation from batch jobs.

Python port of report-generator.ts. Validates job configuration,
calculates date ranges, builds API filters, creates saved report
records, and delegates to the report worker.
"""

import logging
from datetime import datetime, timedelta

from ..services.report_worker import run_report_worker

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Date range calculation
# ---------------------------------------------------------------------------

def calculate_date_range(date_range_type: str) -> dict:
    """Return ``{"date_start": "YYYY-MM-DD", "date_end": "YYYY-MM-DD"}``
    for the given *date_range_type*.

    Supported types:
    - ``last_7_days`` : 7 days ago to yesterday
    - ``last_14_days``: 14 days ago to yesterday
    - ``last_30_days``: 30 days ago to yesterday
    - ``last_week``   : last complete Monday-Sunday
    - ``last_month``  : last complete calendar month
    - default         : falls back to last 7 days
    """
    today = datetime.utcnow().date()
    yesterday = today - timedelta(days=1)

    if date_range_type == "last_7_days":
        date_start = today - timedelta(days=7)
        date_end = yesterday

    elif date_range_type == "last_14_days":
        date_start = today - timedelta(days=14)
        date_end = yesterday

    elif date_range_type == "last_30_days":
        date_start = today - timedelta(days=30)
        date_end = yesterday

    elif date_range_type == "last_week":
        # Last complete Mon-Sun week
        # weekday(): Monday=0 ... Sunday=6
        days_since_monday = today.weekday()  # how far today is from Monday
        last_monday = today - timedelta(days=days_since_monday + 7)
        last_sunday = last_monday + timedelta(days=6)
        date_start = last_monday
        date_end = last_sunday

    elif date_range_type == "last_month":
        # Last complete calendar month
        first_of_this_month = today.replace(day=1)
        last_day_prev_month = first_of_this_month - timedelta(days=1)
        first_of_prev_month = last_day_prev_month.replace(day=1)
        date_start = first_of_prev_month
        date_end = last_day_prev_month

    else:
        # Default: last 7 days
        date_start = today - timedelta(days=7)
        date_end = yesterday

    return {
        "date_start": date_start.isoformat(),
        "date_end": date_end.isoformat(),
    }


# ---------------------------------------------------------------------------
# Report generation
# ---------------------------------------------------------------------------

async def process_report_generation_job(
    job,
    start_time: float,
    db_ops: dict,
) -> None:
    """Process a single report-generation batch job.

    Parameters
    ----------
    job
        A ``BatchJob`` model instance.  Expected attributes:
        ``id``, ``userId``, ``config`` (dict with job parameters).
    start_time : float
        Epoch timestamp when processing started (used for duration tracking).
    db_ops : dict
        A mapping of async database helper functions::

            {
                "update_batch_job": async (job_id, data) -> ...,
                "create_saved_report": async (data) -> saved_report,
                "update_saved_report": async (report_id, data) -> ...,
                "create_batch_history_record": async (data) -> ...,
                "update_batch_history_record": async (record_id, data) -> ...,
                "get_schedule_run": async (run_id) -> ...,
                "update_schedule_run": async (run_id, data) -> ...,
                "get_batch_job": async (job_id) -> ...,
            }
    """

    job_id = job.id
    user_id = job.userId
    config: dict = job.config or {}

    # Unpack db_ops
    update_batch_job = db_ops["update_batch_job"]
    create_saved_report = db_ops["create_saved_report"]
    update_saved_report = db_ops["update_saved_report"]
    create_batch_history_record = db_ops["create_batch_history_record"]
    update_batch_history_record = db_ops["update_batch_history_record"]
    get_schedule_run = db_ops["get_schedule_run"]
    update_schedule_run = db_ops["update_schedule_run"]
    get_batch_job = db_ops["get_batch_job"]

    # ------------------------------------------------------------------
    # Validate required fields
    # ------------------------------------------------------------------
    ad_account_id: str | None = config.get("adAccountId")
    access_token: str | None = config.get("accessToken")

    if not ad_account_id or not access_token:
        error_msg = "Missing required config: adAccountId and accessToken are required"
        logger.error("Job %d validation failed: %s", job_id, error_msg)
        await update_batch_job(
            job_id,
            {"status": "failed", "progress": 0, "error": error_msg},
        )
        return

    # ------------------------------------------------------------------
    # Calculate date range
    # ------------------------------------------------------------------
    date_start: str | None = config.get("dateStart")
    date_end: str | None = config.get("dateEnd")

    if not date_start:
        date_range_type: str = config.get("dateRangeType", "last_7_days")
        date_range = calculate_date_range(date_range_type)
        date_start = date_range["date_start"]
        date_end = date_range["date_end"]
        logger.info(
            "Job %d: calculated date range '%s' -> %s to %s",
            job_id,
            date_range_type,
            date_start,
            date_end,
        )

    level: str = config.get("level") or "account"
    breakdown: str = config.get("breakdown") or "product_id"

    # ------------------------------------------------------------------
    # Filters are applied in Python after CSV download (not at API level)
    # This avoids Facebook API filter issues and allows full data download
    # ------------------------------------------------------------------

    # ------------------------------------------------------------------
    # Create saved_reports record
    # ------------------------------------------------------------------
    report_id = await create_saved_report({
        "userId": user_id,
        "name": f"Report {date_start} to {date_end}",
        "adAccountId": ad_account_id,
        "dateStart": date_start,
        "dateEnd": date_end,
        "level": level,
        "breakdown": breakdown,
        "minSpend": config.get("minSpend"),
        "minCTR": config.get("minCTR"),
        "status": "generating",
        "source": "manual",
    })

    if not report_id:
        raise RuntimeError("Failed to create saved report record")
    logger.info(
        "Job %d: created saved report %d (%s to %s)",
        job_id,
        report_id,
        date_start,
        date_end,
    )

    # ------------------------------------------------------------------
    # Build WorkerConfig
    # ------------------------------------------------------------------
    worker_config: dict = {
        "jobId": job_id,
        "reportId": report_id,
        "userId": user_id,
        "adAccountId": ad_account_id,
        "accessToken": access_token,
        "dateStart": date_start,
        "dateEnd": date_end,
        "level": config.get("level", "account"),
        "breakdown": config.get("breakdown", "product_id"),
        "filters": None,
        "minSpend": float(config["minSpend"]) if config.get("minSpend") else None,
        "minCTR": float(config["minCTR"]) if config.get("minCTR") else None,
        "maxSpend": float(config["maxSpend"]) if config.get("maxSpend") else None,
        "maxCVR": float(config["maxCVR"]) if config.get("maxCVR") else None,
        "topConversionLimit": config.get("topConversionLimit"),
        "updateToCatalog": config.get("updateToCatalog", False),
        "catalogId": config.get("catalogId"),
        "catalogAccessToken": config.get("catalogAccessToken"),
        "customLabel4": config.get("customLabel4"),
        "enableCustomLabel4": config.get("enableCustomLabel4", False),
        "customNumbers": config.get("customNumbers"),
        "customLabels": config.get("customLabels"),
        "scheduleRunId": config.get("scheduleRunId"),
    }

    # ------------------------------------------------------------------
    # Run the worker
    # ------------------------------------------------------------------
    try:
        result = await run_report_worker(
            config=worker_config,
            db_update_job=update_batch_job,
            db_update_report=update_saved_report,
            db_create_history=create_batch_history_record,
            db_update_history=update_batch_history_record,
            db_get_schedule_run=get_schedule_run,
            db_update_schedule_run=update_schedule_run,
            db_get_job=get_batch_job,
        )

        if result.get("success"):
            logger.info(
                "Job %d completed: %d items, $%.2f spend, %dms",
                job_id,
                result.get("totalItems", 0),
                result.get("totalSpend", 0),
                result.get("durationMs", 0),
            )
        else:
            logger.error(
                "Job %d failed: %s (%dms)",
                job_id,
                result.get("error", "unknown error"),
                result.get("durationMs", 0),
            )

    except Exception as exc:
        logger.exception("Job %d raised an unhandled exception: %s", job_id, exc)
        await update_batch_job(
            job_id,
            {"status": "failed", "progress": 0, "error": str(exc)},
        )
        await update_saved_report(
            report_id,
            {"status": "failed", "error": str(exc)},
        )
